AK/Size.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2022, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Concepts.h>
+
+namespace AK {
+
+namespace Detail {
+
+// clang-format off
+template<typename T>
+concept SizeQueryable = requires
+{
+    {
+        declval<AddConst<T>>().size()
+    } -> Unsigned;
+};
+// clang-format on
+
+}
+
+constexpr size_t size(auto begin, auto const& end) requires IteratorPairWith<decltype(begin), decltype(end)>
+{
+    size_t size = 0;
+    while (begin != end) {
```

Comment:
In the STL this is called `distance`. Calling it `size` causes me to be surprised at the O(n) complexity. I guess it should just know because of passing iterators?

Reply:
In the STL getting the number of elements of a set is either (with what I can remember right now) `count_*`, `distance`, `length` or `size` with what ever with the guarantee to be `O(N)` or better depending on the revision of the STL. My guess about `distance` naming would be that they needed to separate iterators parameters because it was a relatively new concepts. And since C++ must be stable, we have to live with these names, that are odd in the present, *forever*...

My personal preference goes to `count`, it can be declined with `count_occurrences`, `count_if`, `count_if_not` and is also also used with enumerations in naming conventions. It has the advantage of being a common prefix for auto completion

Reply:
`std::set::size()` is constant time (O(1)). https://en.cppreference.com/w/cpp/container/set/size

Reply:
`std::list::size` is O(N) before C++11 than O(1) https://en.cppreference.com/w/cpp/container/list/size
Guarantees have changed over time... don't trust a name to guarantee complexity, it just guarantee functionality (modulo the sanity of its implementer).